File: app/src/main/cpp/dsp/biquad_filter.cpp
#include "biquad_filter.h"
#include <algorithm>
#include <cmath>

namespace antigravity {

static constexpr double M_PI_VAL = 3.14159265358979323846;

static inline BiquadCoefficients safeNormalize(double b0, double b1, double b2, double a0, double a1, double a2) {
    if (!std::isfinite(a0) || std::abs(a0) < 1.0e-15) {
        return {1.0, 0.0, 0.0, 0.0, 0.0};
    }
    const double invA0 = 1.0 / a0;
    const double nb0 = b0 * invA0;
    const double nb1 = b1 * invA0;
    const double nb2 = b2 * invA0;
    const double na1 = a1 * invA0;
    const double na2 = a2 * invA0;

    if (!std::isfinite(nb0) || !std::isfinite(nb1) || !std::isfinite(nb2) ||
        !std::isfinite(na1) || !std::isfinite(na2)) {
        return {1.0, 0.0, 0.0, 0.0, 0.0};
    }
    return {nb0, nb1, nb2, na1, na2};
}

BiquadFilter::BiquadFilter() {
    reset();
}

void BiquadFilter::reset() {
    z1_ = 0.0;
    z2_ = 0.0;
    b0_ = 1.0;
    b1_ = 0.0;
    b2_ = 0.0;
    a1_ = 0.0;
    a2_ = 0.0;
}

BiquadCoefficients BiquadFilter::computePeakingEq(double frequency, double q, double gainDb, double sampleRate) {
    if (!std::isfinite(sampleRate) || sampleRate <= 0.0) return {1.0, 0.0, 0.0, 0.0, 0.0};
    frequency = std::clamp(frequency, 1.0, sampleRate * 0.499);
    q = std::clamp(q, 0.01, 100.0);
    gainDb = std::clamp(gainDb, -60.0, 60.0);

    const double A = std::pow(10.0, gainDb / 40.0);
    const double w0 = 2.0 * M_PI_VAL * frequency / sampleRate;
    const double alpha = std::sin(w0) / (2.0 * q);
    const double cosW0 = std::cos(w0);

    const double b0 = 1.0 + alpha * A;
    const double b1 = -2.0 * cosW0;
    const double b2 = 1.0 - alpha * A;
    const double a0 = 1.0 + alpha / A;
    const double a1 = -2.0 * cosW0;
    const double a2 = 1.0 - alpha / A;

    return safeNormalize(b0, b1, b2, a0, a1, a2);
}

BiquadCoefficients BiquadFilter::computeLowShelf(double frequency, double q, double gainDb, double sampleRate) {
    if (!std::isfinite(sampleRate) || sampleRate <= 0.0) return {1.0, 0.0, 0.0, 0.0, 0.0};
    frequency = std::clamp(frequency, 1.0, sampleRate * 0.499);
    q = std::clamp(q, 0.01, 100.0);
    gainDb = std::clamp(gainDb, -60.0, 60.0);

    const double A = std::pow(10.0, gainDb / 40.0);
    const double w0 = 2.0 * M_PI_VAL * frequency / sampleRate;
    const double cosW0 = std::cos(w0);
    const double sinW0 = std::sin(w0);
    const double alpha = sinW0 / (2.0 * q);
    const double sqrtA = std::sqrt(A);

    const double b0 = A * ((A + 1.0) - (A - 1.0) * cosW0 + 2.0 * sqrtA * alpha);
    const double b1 = 2.0 * A * ((A - 1.0) - (A + 1.0) * cosW0);
    const double b2 = A * ((A + 1.0) - (A - 1.0) * cosW0 - 2.0 * sqrtA * alpha);
    const double a0 = (A + 1.0) + (A - 1.0) * cosW0 + 2.0 * sqrtA * alpha;
    const double a1 = -2.0 * ((A - 1.0) + (A + 1.0) * cosW0);
    const double a2 = (A + 1.0) + (A - 1.0) * cosW0 - 2.0 * sqrtA * alpha;

    return safeNormalize(b0, b1, b2, a0, a1, a2);
}

BiquadCoefficients BiquadFilter::computeHighShelf(double frequency, double q, double gainDb, double sampleRate) {
    if (!std::isfinite(sampleRate) || sampleRate <= 0.0) return {1.0, 0.0, 0.0, 0.0, 0.0};
    frequency = std::clamp(frequency, 1.0, sampleRate * 0.499);
    q = std::clamp(q, 0.01, 100.0);
    gainDb = std::clamp(gainDb, -60.0, 60.0);

    const double A = std::pow(10.0, gainDb / 40.0);
    const double w0 = 2.0 * M_PI_VAL * frequency / sampleRate;
    const double cosW0 = std::cos(w0);
    const double sinW0 = std::sin(w0);
    const double alpha = sinW0 / (2.0 * q);
    const double sqrtA = std::sqrt(A);

    const double b0 = A * ((A + 1.0) + (A - 1.0) * cosW0 + 2.0 * sqrtA * alpha);
    const double b1 = -2.0 * A * ((A - 1.0) + (A + 1.0) * cosW0);
    const double b2 = A * ((A + 1.0) + (A - 1.0) * cosW0 - 2.0 * sqrtA * alpha);
    const double a0 = (A + 1.0) - (A - 1.0) * cosW0 + 2.0 * sqrtA * alpha;
    const double a1 = 2.0 * ((A - 1.0) - (A + 1.0) * cosW0);
    const double a2 = (A + 1.0) - (A - 1.0) * cosW0 - 2.0 * sqrtA * alpha;

    return safeNormalize(b0, b1, b2, a0, a1, a2);
}

BiquadCoefficients BiquadFilter::computeLowPass(double frequency, double q, double sampleRate) {
    if (!std::isfinite(sampleRate) || sampleRate <= 0.0) return {1.0, 0.0, 0.0, 0.0, 0.0};
    frequency = std::clamp(frequency, 1.0, sampleRate * 0.499);
    q = std::clamp(q, 0.01, 100.0);

    const double w0 = 2.0 * M_PI_VAL * frequency / sampleRate;
    const double cosW0 = std::cos(w0);
    const double alpha = std::sin(w0) / (2.0 * q);

    const double b0 = (1.0 - cosW0) * 0.5;
    const double b1 = 1.0 - cosW0;
    const double b2 = (1.0 - cosW0) * 0.5;
    const double a0 = 1.0 + alpha;
    const double a1 = -2.0 * cosW0;
    const double a2 = 1.0 - alpha;

    return safeNormalize(b0, b1, b2, a0, a1, a2);
}

BiquadCoefficients BiquadFilter::computeHighPass(double frequency, double q, double sampleRate) {
    if (!std::isfinite(sampleRate) || sampleRate <= 0.0) return {1.0, 0.0, 0.0, 0.0, 0.0};
    frequency = std::clamp(frequency, 1.0, sampleRate * 0.499);
    q = std::clamp(q, 0.01, 100.0);

    const double w0 = 2.0 * M_PI_VAL * frequency / sampleRate;
    const double cosW0 = std::cos(w0);
    const double alpha = std::sin(w0) / (2.0 * q);

    const double b0 = (1.0 + cosW0) * 0.5;
    const double b1 = -(1.0 + cosW0);
    const double b2 = (1.0 + cosW0) * 0.5;
    const double a0 = 1.0 + alpha;
    const double a1 = -2.0 * cosW0;
    const double a2 = 1.0 - alpha;

    return safeNormalize(b0, b1, b2, a0, a1, a2);
}

BiquadCoefficients BiquadFilter::computeBandPass(double frequency, double q, double sampleRate) {
    if (!std::isfinite(sampleRate) || sampleRate <= 0.0) return {1.0, 0.0, 0.0, 0.0, 0.0};
    frequency = std::clamp(frequency, 1.0, sampleRate * 0.499);
    q = std::clamp(q, 0.01, 100.0);

    const double w0 = 2.0 * M_PI_VAL * frequency / sampleRate;
    const double cosW0 = std::cos(w0);
    const double alpha = std::sin(w0) / (2.0 * q);

    const double b0 = alpha;
    const double b1 = 0.0;
    const double b2 = -alpha;
    const double a0 = 1.0 + alpha;
    const double a1 = -2.0 * cosW0;
    const double a2 = 1.0 - alpha;

    return safeNormalize(b0, b1, b2, a0, a1, a2);
}

BiquadCoefficients BiquadFilter::computeNotch(double frequency, double q, double sampleRate) {
    if (!std::isfinite(sampleRate) || sampleRate <= 0.0) return {1.0, 0.0, 0.0, 0.0, 0.0};
    frequency = std::clamp(frequency, 1.0, sampleRate * 0.499);
    q = std::clamp(q, 0.01, 100.0);

    const double w0 = 2.0 * M_PI_VAL * frequency / sampleRate;
    const double cosW0 = std::cos(w0);
    const double alpha = std::sin(w0) / (2.0 * q);

    const double b0 = 1.0;
    const double b1 = -2.0 * cosW0;
    const double b2 = 1.0;
    const double a0 = 1.0 + alpha;
    const double a1 = -2.0 * cosW0;
    const double a2 = 1.0 - alpha;

    return safeNormalize(b0, b1, b2, a0, a1, a2);
}

BiquadCoefficients BiquadFilter::computeAllPass(double frequency, double q, double sampleRate) {
    if (!std::isfinite(sampleRate) || sampleRate <= 0.0) return {1.0, 0.0, 0.0, 0.0, 0.0};
    frequency = std::clamp(frequency, 1.0, sampleRate * 0.499);
    q = std::clamp(q, 0.01, 100.0);

    const double w0 = 2.0 * M_PI_VAL * frequency / sampleRate;
    const double cosW0 = std::cos(w0);
    const double alpha = std::sin(w0) / (2.0 * q);

    const double b0 = 1.0 - alpha;
    const double b1 = -2.0 * cosW0;
    const double b2 = 1.0 + alpha;
    const double a0 = 1.0 + alpha;
    const double a1 = -2.0 * cosW0;
    const double a2 = 1.0 - alpha;

    return safeNormalize(b0, b1, b2, a0, a1, a2);
}

void BiquadFilter::setPeakingEq(double frequency, double q, double gainDb, double sampleRate) {
    setCoefficients(computePeakingEq(frequency, q, gainDb, sampleRate));
}

void BiquadFilter::setLowShelf(double frequency, double q, double gainDb, double sampleRate) {
    setCoefficients(computeLowShelf(frequency, q, gainDb, sampleRate));
}

void BiquadFilter::setHighShelf(double frequency, double q, double gainDb, double sampleRate) {
    setCoefficients(computeHighShelf(frequency, q, gainDb, sampleRate));
}

void BiquadFilter::setLowPass(double frequency, double q, double sampleRate) {
    setCoefficients(computeLowPass(frequency, q, sampleRate));
}

void BiquadFilter::setHighPass(double frequency, double q, double sampleRate) {
    setCoefficients(computeHighPass(frequency, q, sampleRate));
}

void BiquadFilter::setBandPass(double frequency, double q, double sampleRate) {
    setCoefficients(computeBandPass(frequency, q, sampleRate));
}

void BiquadFilter::setNotch(double frequency, double q, double sampleRate) {
    setCoefficients(computeNotch(frequency, q, sampleRate));
}

void BiquadFilter::setAllPass(double frequency, double q, double sampleRate) {
    setCoefficients(computeAllPass(frequency, q, sampleRate));
}

} // namespace antigravity

File: app/src/main/java/com/tensorix/antigravityplayer/audio/BiquadFilter.kt
package com.tensorix.antigravityplayer.audio

import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

data class BiquadCoeffs(
    val b0: Double = 1.0,
    val b1: Double = 0.0,
    val b2: Double = 0.0,
    val a1: Double = 0.0,
    val a2: Double = 0.0
)

/**
 * 64-bit Double Precision Direct Form II Transposed Biquad Filter
 * Numerical stability for audiophile parametric equalization without rounding noise.
 */
class BiquadFilter {
    private var b0 = 1.0
    private var b1 = 0.0
    private var b2 = 0.0
    private var a1 = 0.0
    private var a2 = 0.0

    private var z1 = 0.0
    private var z2 = 0.0

    fun reset() {
        z1 = 0.0
        z2 = 0.0
    }

    fun setCoefficients(c: BiquadCoeffs) {
        b0 = c.b0
        b1 = c.b1
        b2 = c.b2
        a1 = c.a1
        a2 = c.a2
    }

    fun getCoefficients(): BiquadCoeffs = BiquadCoeffs(b0, b1, b2, a1, a2)

    fun process(x: Double): Double {
        val y = b0 * x + z1
        z1 = b1 * x - a1 * y + z2
        z2 = b2 * x - a2 * y
        // Audiophile denormal protection against CPU Kryo stalls at ultra-low amplitudes
        if (z1 != 0.0 && abs(z1) < 1.0e-15) z1 = 0.0
        if (z2 != 0.0 && abs(z2) < 1.0e-15) z2 = 0.0
        return y
    }

    fun setPeakingEq(f0: Double, q: Double, gainDb: Double, fs: Double) {
        setCoefficients(computePeakingEq(f0, q, gainDb, fs))
    }

    fun setLowShelf(f0: Double, q: Double, gainDb: Double, fs: Double) {
        setCoefficients(computeLowShelf(f0, q, gainDb, fs))
    }

    fun setHighShelf(f0: Double, q: Double, gainDb: Double, fs: Double) {
        setCoefficients(computeHighShelf(f0, q, gainDb, fs))
    }

    fun setHighPass(f0: Double, q: Double, fs: Double) {
        setCoefficients(computeHighPass(f0, q, fs))
    }

    fun setLowPass(f0: Double, q: Double, fs: Double) {
        setCoefficients(computeLowPass(f0, q, fs))
    }

    fun setAllPass(f0: Double, q: Double, fs: Double) {
        setCoefficients(computeAllPass(f0, q, fs))
    }

    fun setNotch(f0: Double, q: Double, fs: Double) {
        setCoefficients(computeNotch(f0, q, fs))
    }

    companion object {
        private fun safeNormalize(b0: Double, b1: Double, b2: Double, a0: Double, a1: Double, a2: Double): BiquadCoeffs {
            if (!a0.isFinite() || abs(a0) < 1.0e-15) return BiquadCoeffs()
            val invA0 = 1.0 / a0
            val nb0 = b0 * invA0
            val nb1 = b1 * invA0
            val nb2 = b2 * invA0
            val na1 = a1 * invA0
            val na2 = a2 * invA0
            if (!nb0.isFinite() || !nb1.isFinite() || !nb2.isFinite() || !na1.isFinite() || !na2.isFinite()) {
                return BiquadCoeffs()
            }
            return BiquadCoeffs(nb0, nb1, nb2, na1, na2)
        }

        fun computePeakingEq(f0: Double, q: Double, gainDb: Double, fs: Double): BiquadCoeffs {
            if (!fs.isFinite() || fs <= 0.0) return BiquadCoeffs()
            val f = f0.coerceIn(1.0, fs * 0.499)
            val qSafe = q.coerceIn(0.01, 100.0)
            val g = gainDb.coerceIn(-60.0, 60.0)
            if (abs(g) < 1.0e-6) return BiquadCoeffs()

            val a = 10.0.pow(g / 40.0)
            val w0 = 2.0 * Math.PI * f / fs
            val alpha = sin(w0) / (2.0 * qSafe)
            val cosW0 = cos(w0)

            val a0 = 1.0 + alpha / a
            val b0 = 1.0 + alpha * a
            val b1 = -2.0 * cosW0
            val b2 = 1.0 - alpha * a
            val a1 = -2.0 * cosW0
            val a2 = 1.0 - alpha / a
            return safeNormalize(b0, b1, b2, a0, a1, a2)
        }

        fun computeLowShelf(f0: Double, q: Double, gainDb: Double, fs: Double): BiquadCoeffs {
            if (!fs.isFinite() || fs <= 0.0) return BiquadCoeffs()
            val f = f0.coerceIn(1.0, fs * 0.499)
            val qSafe = q.coerceIn(0.01, 100.0)
            val g = gainDb.coerceIn(-60.0, 60.0)
            if (abs(g) < 1.0e-6) return BiquadCoeffs()

            val a = 10.0.pow(g / 40.0)
            val w0 = 2.0 * Math.PI * f / fs
            val alpha = sin(w0) / 2.0 * sqrt((a + 1.0 / a) * (1.0 / qSafe - 1.0) + 2.0)
            val cosW0 = cos(w0)

            val a0 = (a + 1.0) + (a - 1.0) * cosW0 + 2.0 * sqrt(a) * alpha
            val b0 = a * ((a + 1.0) - (a - 1.0) * cosW0 + 2.0 * sqrt(a) * alpha)
            val b1 = 2.0 * a * ((a - 1.0) - (a + 1.0) * cosW0)
            val b2 = a * ((a + 1.0) - (a - 1.0) * cosW0 - 2.0 * sqrt(a) * alpha)
            val a1 = -2.0 * ((a - 1.0) + (a + 1.0) * cosW0)
            val a2 = (a + 1.0) + (a - 1.0) * cosW0 - 2.0 * sqrt(a) * alpha
            return safeNormalize(b0, b1, b2, a0, a1, a2)
        }

        fun computeHighShelf(f0: Double, q: Double, gainDb: Double, fs: Double): BiquadCoeffs {
            if (!fs.isFinite() || fs <= 0.0) return BiquadCoeffs()
            val f = f0.coerceIn(1.0, fs * 0.499)
            val qSafe = q.coerceIn(0.01, 100.0)
            val g = gainDb.coerceIn(-60.0, 60.0)
            if (abs(g) < 1.0e-6) return BiquadCoeffs()

            val a = 10.0.pow(g / 40.0)
            val w0 = 2.0 * Math.PI * f / fs
            val alpha = sin(w0) / 2.0 * sqrt((a + 1.0 / a) * (1.0 / qSafe - 1.0) + 2.0)
            val cosW0 = cos(w0)

            val a0 = (a + 1.0) - (a - 1.0) * cosW0 + 2.0 * sqrt(a) * alpha
            val b0 = a * ((a + 1.0) + (a - 1.0) * cosW0 + 2.0 * sqrt(a) * alpha)
            val b1 = -2.0 * a * ((a - 1.0) + (a + 1.0) * cosW0)
            val b2 = a * ((a + 1.0) - (a - 1.0) * cosW0 - 2.0 * sqrt(a) * alpha)
            val a1 = 2.0 * ((a - 1.0) - (a + 1.0) * cosW0)
            val a2 = (a + 1.0) - (a - 1.0) * cosW0 - 2.0 * sqrt(a) * alpha
            return safeNormalize(b0, b1, b2, a0, a1, a2)
        }

        fun computeHighPass(f0: Double, q: Double, fs: Double): BiquadCoeffs {
            if (!fs.isFinite() || fs <= 0.0) return BiquadCoeffs()
            val f = f0.coerceIn(1.0, fs * 0.499)
            val qSafe = q.coerceIn(0.01, 100.0)
            val w0 = 2.0 * Math.PI * f / fs
            val alpha = sin(w0) / (2.0 * qSafe)
            val cosW0 = cos(w0)

            val a0 = 1.0 + alpha
            val b0 = (1.0 + cosW0) * 0.5
            val b1 = -(1.0 + cosW0)
            val b2 = (1.0 + cosW0) * 0.5
            val a1 = -2.0 * cosW0
            val a2 = 1.0 - alpha
            return safeNormalize(b0, b1, b2, a0, a1, a2)
        }

        fun computeLowPass(f0: Double, q: Double, fs: Double): BiquadCoeffs {
            if (!fs.isFinite() || fs <= 0.0) return BiquadCoeffs()
            val f = f0.coerceIn(1.0, fs * 0.499)
            val qSafe = q.coerceIn(0.01, 100.0)
            val w0 = 2.0 * Math.PI * f / fs
            val alpha = sin(w0) / (2.0 * qSafe)
            val cosW0 = cos(w0)

            val a0 = 1.0 + alpha
            val b0 = (1.0 - cosW0) * 0.5
            val b1 = 1.0 - cosW0
            val b2 = (1.0 - cosW0) * 0.5
            val a1 = -2.0 * cosW0
            val a2 = 1.0 - alpha
            return safeNormalize(b0, b1, b2, a0, a1, a2)
        }

        fun computeAllPass(f0: Double, q: Double, fs: Double): BiquadCoeffs {
            if (!fs.isFinite() || fs <= 0.0) return BiquadCoeffs()
            val f = f0.coerceIn(1.0, fs * 0.499)
            val qSafe = q.coerceIn(0.01, 100.0)
            val w0 = 2.0 * Math.PI * f / fs
            val alpha = sin(w0) / (2.0 * qSafe)
            val cosW0 = cos(w0)

            val a0 = 1.0 + alpha
            val b0 = 1.0 - alpha
            val b1 = -2.0 * cosW0
            val b2 = 1.0 + alpha
            val a1 = -2.0 * cosW0
            val a2 = 1.0 - alpha
            return safeNormalize(b0, b1, b2, a0, a1, a2)
        }

        fun computeNotch(f0: Double, q: Double, fs: Double): BiquadCoeffs {
            if (!fs.isFinite() || fs <= 0.0) return BiquadCoeffs()
            val f = f0.coerceIn(1.0, fs * 0.499)
            val qSafe = q.coerceIn(0.01, 100.0)
            val w0 = 2.0 * Math.PI * f / fs
            val alpha = sin(w0) / (2.0 * qSafe)
            val cosW0 = cos(w0)

            val a0 = 1.0 + alpha
            val b0 = 1.0
            val b1 = -2.0 * cosW0
            val b2 = 1.0
            val a1 = -2.0 * cosW0
            val a2 = 1.0 - alpha
            return safeNormalize(b0, b1, b2, a0, a1, a2)
        }
    }
}
